program/rust/src/rust_oracle.rs

```diff
@@ -32,47 +40,154 @@ use crate::c_oracle_header::{
 ) -> OracleResult {
-    //For now, we did not change the behavior of this. this is just to show the proposed structure
-    // of the program
-    c_entrypoint_wrapper(input)
+    let price_account_info = &accounts[1];
+    let account_len = price_account_info.try_data_len()?;
+    if account_len < PRICE_ACCOUNT_SIZE {
+        if account_len != PRICE_T_SIZE {
+            return Err(ProgramError::InvalidArgument);
+        }
+        msg!("Please resize the account to allow for SMA tracking!");
+        return c_entrypoint_wrapper(input);
+    }
+    let c_ret_value = c_entrypoint_wrapper(input)?;
+    if c_ret_value == SUCCESSFULLY_UPDATED_AGGREGATE {
+        let mut price_account = load_account_as_mut::<PriceAccountWrapper>(price_account_info)?;
+        price_account.add_price_to_time_machine()?;
+        msg!("updated tracker!");
+    }
+    Ok(c_ret_value)
+}
+fn send_lamports<'a>(
+    from: &AccountInfo<'a>,
+    to: &AccountInfo<'a>,
+    system_program: &AccountInfo<'a>,
+    ammount: u64,
+) -> Result<(), ProgramError> {
+    let transfer_instruction = transfer(from.key, to.key, ammount);
+    invoke(
+        &transfer_instruction,
+        &[from.clone(), to.clone(), system_program.clone()],
+    )?;
+    Ok(())
+}
+
+///resize price account and initialize the TimeMachineStructure
+pub fn upgrade_price_account<'a>(
+    funding_account_info: &AccountInfo<'a>,
+    price_account_info: &AccountInfo<'a>,
+    system_program: &AccountInfo<'a>,
+    program_id: &Pubkey,
+) -> OracleResult {
+    check_valid_funding_account(funding_account_info)?;
+    check_valid_signable_account(program_id, price_account_info, size_of::<pc_price_t>())?;
+    let account_len = price_account_info.try_data_len()?;
+    match account_len {
+        PRICE_T_SIZE => {
+            //ensure account is still rent exempt after resizing
+            let rent: Rent = Default::default();
+            let lamports_needed: u64 = rent
+                .minimum_balance(size_of::<PriceAccountWrapper>())
+                .saturating_sub(price_account_info.lamports());
+            if lamports_needed > 0 {
+                send_lamports(
+                    funding_account_info,
+                    price_account_info,
+                    system_program,
+                    lamports_needed,
+                )?;
+            }
+            //resize
+            //we do not need to zero initialize since this is the first time this memory
+            //is allocated
+            price_account_info.realloc(size_of::<PriceAccountWrapper>(), false)?;
+            //update twap_tracker
+            let mut price_account = load_account_as_mut::<PriceAccountWrapper>(price_account_info)?;
+            price_account.invalidate_current_time_machine_entries()?;
+            Ok(SUCCESS)
+        }
+        PRICE_ACCOUNT_SIZE => Ok(SUCCESS),
+        _ => Err(ProgramError::InvalidArgument),
+    }
+}
+
+fn get_account_type(account: &AccountInfo) -> Result<u32, ProgramError> {
```

Comment:
nitpick: this doesn't seem like it's worth a helper function -- it's the same number of lines of code to write the one line body inline.